### DISCUSSION TITLE
Question about mac catalyst corner radius

### DISCUSSION THREAD

**Discussion by iDevelopper** [LeoNatan/LNPopupController#488] [General] [answered]
Hello Leo,

Can I ask you where and/or how to find information about the windows corner radius used on macOS (via mac catalyst) ?

May be a private api method.

I allow myself to ask you this question because I can find absolutely nothing on the subject.

Thanks.

**Comment by LeoNatan** [comment]
Hello, do you mean the macOS window corner radius? That is the default look that `NSWindow` has. If you want a custom shape, you can check [this old example](https://developer.apple.com/library/archive/samplecode/RoundTransparentWindow/Introduction/Intro.html), but accessing the underlying `NSWindow` is not simple, and is private API. You can see how I obtain the `NSWindow` here: https://github.com/LeoNatan/LNPopupController/blob/8fb8edde74bb2e45d8a9b5ddd539a0b8af82b904/LNPopupController/LNPopupController/Private/UIView%2BLNPopupSupportPrivate.m#L225

**Reply by iDevelopper** [reply]
First, thank you very much for your time and your answer.

> do you mean the macOS window corner radius?

Yes this is what I want to know. I can't find a default value for the macOS windows as I can find the screen corner radius of an iPhone/iPad screen for iOS with the code below:


```swift
import UIKit

//_displayCornerRadius
private let _dCR = "X2Rpc3BsYXlDb3JuZXJSYWRpdXM="

extension UIScreen {
    
    private static let cornerRadiusKey: String = {
        return decodeBase64String(base64String: _dCR)!
    }()

    /// The corner radius of the display. Uses a private property of `UIScreen`,
    /// and may report 0 if the API changes.
    public var displayCornerRadius: CGFloat {
        guard let cornerRadius = self.value(forKey: Self.cornerRadiusKey) as? CGFloat else {
            return 0
        }

        return cornerRadius
    }
}
```

39.0 for an iPhone 11 Pro

**Reply by LeoNatan** [reply]
You can get the Mac window corner radius like this:

```objc
id hostingWindow = [self.view.window valueForKeyPath:@"hostWindow.attachedWindow"];
id themeFrame = [hostingWindow valueForKey:@"themeFrame"];
CGFloat windowCornerRadius = [[themeFrame valueForKey:@"bottomCornerHeight"] doubleValue];
```

See example project:
[WindowCornerRadiusTest.zip](https://github.com/LeoNatan/LNPopupController/files/7453718/WindowCornerRadiusTest.zip)

**Reply by LeoNatan** [reply]
@iDevelopper Did it work for you?

**Reply by iDevelopper** [reply]
Hi Leo, I'm just testing, it's really neat. I don't know how to thank you.
However, I don't understand why on a break point, looking at the detail of the hostingWindow variable, I don't see the "themeFrame" variable. I see a variable named "_borderView" (which appears to be the same object).

<img width="1017" alt="image" src="https://user-images.githubusercontent.com/1782450/139905948-f243ee7c-72fa-4c9b-a7a0-8b715a3a3c09.png">

Then I can't find this "bottomCornerHeight" variable when expanding the borderView ...

<img width="1026" alt="image" src="https://user-images.githubusercontent.com/1782450/139906602-ac224f54-8999-46b3-963d-b88b4216620c.png">

**Reply by LeoNatan** [reply]
Those are just the instance variables (ivar) that are exposed. If you want to find the private ivars or methods/properties, you should use `fp_methodDescription` and `fp_ivarDescription` in the debugger like so:
```
po [hostingWindow fp_methodDescription]
```
of in debugger in Swift context:
```
po hostingWindow.value(forKey: "fp_methodDescription")
```

**Reply by iDevelopper** [reply]
<img width="1145" alt="image" src="https://user-images.githubusercontent.com/1782450/139912284-c7b27100-dae6-491a-874d-20bc51020ada.png">

**Reply by LeoNatan** [reply]
Sorry, no idea why it isn’t working for you.

**Reply by iDevelopper** [reply, marked as answer]
If it works for you with the same sample project and the same version of Xcode (13.0?), It will work for me. I will look. Anyway, thank you for all 👍☺️

**Reply by LeoNatan** [reply]
Yes, it worked for me. That’s how I found those private keys.
